File: routes/admin/goodsadd.js
var res = require('./lib/resp.js');
var cates = require("./models/cates.js");
var goods = require("./models/goods.js");
var filepath = require('path/filepath');
var dbUtils = require('./models/dbUtils.js');
var io = require('io');
var os = require('os');
var strings = require('strings');
var Dirs = require("./lib/Dirs.js");
var utils = require('./lib/utils.js');


if (!session.get("isAdmin")) {
  return res.redirect('/routes/admin/login.js');
}

if (request.getMethod() === 'GET') {
  var allData = cates.find();
  var data = {
    tree: cates.getTree(allData),
  };
  return res.render('./views/admin/goodsadd.ejs', data);
}

if (request.getMethod() === 'POST') {
  var err = request.parseMultipartForm(1024 * 1024)
  if (err) {
    throw err
  }

  var data = {
    oid: dbUtils.makeOid(),
    goods_name: request.formValue('goods_name'),
    cat_id: request.formValue('cat_id'),
    shop_price: request.formValue('shop_price'),
    goods_desc: request.formValue('goods_desc'),
    goods_number: utils.toInt(request.formValue('goods_number'), 0),
    is_best: utils.toInt(request.formValue('is_best'), 0),
    is_new: utils.toInt(request.formValue('is_new'), 0),
    is_hot: utils.toInt(request.formValue('is_hot'), 0),
    is_on_sale: utils.toInt(request.formValue('is_on_sale'), 0),
  };

  var o = request.formFile('goods_img');
  if (o.err) {
    if (request.isMissingFile(o.err)) {
      res.send("未上传图片");
      return;
    }
    throw o.err;
  }
  var contentType = o.header['Content-Type'][0];
  if (!strings.hasPrefix(contentType, "image")) {
    res.send("上传的不是图片");
    return;
  }
  var ext = filepath.ext(o.name);
  var oidFilename = dbUtils.makeOid() + ext;
  var dateDir = Dirs.getDateDir();
  // 文件的本地磁盘路径
  var filename = filepath.join(dateDir.fullpath, oidFilename);
  // web访问路径
  var publicPath = dateDir.dir + oidFilename;

  var r = os.openFile(filename, os.O_CREATE | os.O_WRONLY, 0666)
  if (r.err) {
    throw r.err;
  }
  var file = r.value
  r = io.copy(file, o.file)
  if (r.err) {
    throw r.err;
  }
  o.file.close()
  file.close()

  data.goods_img = publicPath;
  // res.json(data)
  console.log("%j", data)
  goods.add(data);

  res.redirect('./goodslist.js');
}

File: routes/admin/goodsadd.go
package admin

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/zengming00/go-testShop/framework"
	"github.com/zengming00/go-testShop/lib"
	"github.com/zengming00/go-testShop/models"
)

func GoodsAdd(ctx *framework.HandlerContext) {
	var isAdmin = false
	if v, ok := ctx.GetSessionVal("isAdmin"); ok {
		isAdmin = v.(bool)
	}
	if !isAdmin {
		ctx.Redirect("/routes/admin/login.go")
	}

	if ctx.R.Method == "GET" {
		var allData = ctx.Cates.Find()
		var data = map[string]interface{}{
			"tree": ctx.Cates.GetTree(allData),
		}
		ctx.Render("./views/admin/goodsadd.html", data, lib.FuncMap)
		return
	}

	if ctx.R.Method == "POST" {
		var err = ctx.R.ParseMultipartForm(1024 * 1024)
		if err != nil {
			panic(err)
		}

		formFile, fh, err := ctx.R.FormFile("goods_img")
		if err == http.ErrMissingFile {
			ctx.Send("未上传图片")
			return
		}
		if err != nil {
			panic(err)
		}
		defer formFile.Close()

		var contentType = fh.Header.Get("Content-Type")
		if !strings.HasPrefix(contentType, "image") {
			ctx.Send("上传的不是图片")
			return
		}
		var ext = filepath.Ext(fh.Filename)
		var oidFilename = models.MakeOid() + ext
		var dateDir = lib.GetDateDir()
		// 文件的本地磁盘路径
		var filename = filepath.Join(dateDir.Fullpath, oidFilename)
		// web访问路径
		var publicPath = dateDir.Dir + oidFilename

		file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0666)
		if err != nil {
			panic(err)
		}
		defer file.Close()

		_, err = io.Copy(file, formFile)
		if err != nil {
			panic(err)
		}

		var data = map[string]interface{}{
			"oid":          models.MakeOid(),
			"goods_name":   ctx.R.FormValue("goods_name"),
			"cat_id":       ctx.R.FormValue("cat_id"),
			"shop_price":   ctx.R.FormValue("shop_price"),
			"goods_desc":   ctx.R.FormValue("goods_desc"),
			"goods_number": lib.ParsePositiveInt(ctx.R.FormValue("goods_number"), 0),
			"is_best":      lib.ParsePositiveInt(ctx.R.FormValue("is_best"), 0),
			"is_new":       lib.ParsePositiveInt(ctx.R.FormValue("is_new"), 0),
			"is_hot":       lib.ParsePositiveInt(ctx.R.FormValue("is_hot"), 0),
			"is_on_sale":   lib.ParsePositiveInt(ctx.R.FormValue("is_on_sale"), 0),
			"goods_img":    publicPath,
		}

		var result = ctx.Goods.Add(data)
		fmt.Printf("%#v\n", result)
		// ctx.Json(data)
		ctx.Redirect("./goodslist.go")
	}
}
